src/pudl/transform/eia861.py

```diff
@@ -2327,76 +2146,290 @@ def utility_data(tfr_dfs):
-        "operational_data_eia861": operational_data,
-        "reliability_eia861": reliability,
-        "utility_data_eia861": utility_data,
-    }
+    # The dataframes from which to compile BA-Util-State associations
+    ba_eia861 = dfs.pop("clean_balancing_authority_eia861")
+    data_dfs = list(dfs.values())
 
-    # Dictionary for transformed dataframes and pre-transformed dataframes.
-    # Pre-transformed dataframes may be split into two or more output dataframes.
-    tfr_dfs = {}
+    logger.info("Building an EIA 861 BA-Util-State association table.")
 
-    if not raw_dfs:
-        logger.info("No raw EIA 861 dataframes found. Not transforming EIA 861.")
-        return tfr_dfs
-    # Run each of the requested transform funtctions
-    for tfr_func in eia861_settings.transform_functions:
-        logger.info(
-            f"Transforming raw EIA 861 DataFrames for {tfr_func} "
-            f"concatenated across all years."
+    # Helpful shorthand query strings....
+    early_years = "report_date<='2012-12-31'"
+    late_years = "report_date>='2013-01-01'"
+    early_dfs = [df.query(early_years) for df in data_dfs]
+    late_dfs = [df.query(late_years) for df in data_dfs]
+
+    # The old BA table lists utilities directly, but has no state information.
+    early_date_ba_util = _harvest_associations(
+        dfs=[
+            ba_eia861.query(early_years),
+        ],
+        cols=["report_date", "balancing_authority_id_eia", "utility_id_eia"],
+    )
+    # State-utility associations are brought in from observations in data_dfs
+    early_date_util_state = _harvest_associations(
+        dfs=early_dfs,
+        cols=["report_date", "utility_id_eia", "state"],
+    )
+    early_date_ba_util_state = early_date_ba_util.merge(
+        early_date_util_state, how="outer"
+    ).drop_duplicates()
+
+    # New BA table has no utility information, but has BA Codes...
+    late_ba_code_id = _harvest_associations(
+        dfs=[
+            ba_eia861.query(late_years),
+        ],
+        cols=[
+            "report_date",
+            "balancing_authority_code_eia",
+            "balancing_authority_id_eia",
+        ],
+    )
+    # BA Code allows us to bring in utility+state data from data_dfs:
+    late_date_ba_code_util_state = _harvest_associations(
+        dfs=late_dfs,
+        cols=["report_date", "balancing_authority_code_eia", "utility_id_eia", "state"],
+    )
+    # We merge on ba_code then drop it, b/c only BA ID exists in all years consistently:
+    late_date_ba_util_state = (
+        late_date_ba_code_util_state.merge(late_ba_code_id, how="outer")
+        .drop("balancing_authority_code_eia", axis="columns")
+        .drop_duplicates()
+    )
+
+    ba_assn_eia861 = (
+        pd.concat([early_date_ba_util_state, late_date_ba_util_state])
+        .dropna(subset=["balancing_authority_id_eia"])
+        .pipe(apply_pudl_dtypes, group="eia")
+    )
+    return ba_assn_eia861
+
+
+@asset
+def balancing_authority_eia861(
```

Comment:
Moved this function down from above, to group it with the other "secondary transform" functions.